firedrake/preconditioners/fdm.py

```diff
@@ -0,0 +1,1036 @@
+
+        self.assemble_kron(prealloc, V, Gq, Bq, Afdm, Dfdm, eta, bcflags, needs_interior_facet)
+        nnz = get_preallocation(prealloc, ndof)
+        self.Pmat = PETSc.Mat().createAIJ(A.getSizes(), nnz=nnz, comm=A.comm)
+        self._assemble_Pmat = partial(self.assemble_kron, self.Pmat, V, Gq, Bq,
+                                      Afdm, Dfdm, eta, bcflags, needs_interior_facet)
+
+        prealloc.destroy()
+        lgmap = V.dof_dset.lgmap
+        self.Pmat.setBlockSize(V.value_size)
+        self.Pmat.setLGMap(lgmap, lgmap)
+
+        opc = pc
+        # Internally, we just set up a PC object that the user can configure
+        # however from the PETSc command line.  Since PC allows the user to specify
+        # a KSP, we can do iterative by -fdm_pc_type ksp.
+        pc = PETSc.PC().create(comm=opc.comm)
+        pc.incrementTabLevel(1, parent=opc)
+
+        # We set a DM on the constructed PC so one
+        # can do patch solves with ASMPC.
+        from firedrake.solving_utils import _SNESContext
+        mat_type = "aij"
+        dm = opc.getDM()
+        octx = get_appctx(dm)
+        oproblem = octx._problem
+        self._ctx_ref = _SNESContext(oproblem, mat_type, mat_type, octx.appctx, options_prefix=options_prefix)
+
+        pc.setDM(dm)
+        pc.setOptionsPrefix(options_prefix)
+        pc.setOperators(self.Pmat, self.Pmat)
+        self.pc = pc
+        with dmhooks.add_hooks(dm, self, appctx=self._ctx_ref, save=False):
+            pc.setFromOptions()
+        self.update(pc)
+
+    def update(self, pc):
+        self._assemble_Gq()
+        self._assemble_Bq()
+        self.Pmat.zeroEntries()
+        self._assemble_Pmat()
+        self.Pmat.zeroRowsColumnsLocal(self.bc_nodes)
+
+    def applyTranspose(self, pc, x, y):
+        # TODO trivial to implement reusing the code below
+        pass
+
+    def apply(self, pc, x, y):
+        self.uc.assign(firedrake.zero())
+
+        with self.uf.dat.vec_wo as xf:
+            x.copy(xf)
+
+        op2.par_loop(self.restrict_kernel, self.mesh.cell_set,
+                     self.uc.dat(op2.INC, self.cell_node_map),
+                     self.uf.dat(op2.READ, self.cell_node_map),
+                     self.weight.dat(op2.READ, self.cell_node_map))
+
+        for bc in self.bcs:
+            bc.zero(self.uc)
+
+        dm = pc.getDM()
+        with dmhooks.add_hooks(dm, self, appctx=self._ctx_ref), self.uc.dat.vec as x_, self.uf.dat.vec as y_:
+            self.pc.apply(x_, y_)
+
+        for bc in self.bcs:
+            bc.zero(self.uf)
+
+        op2.par_loop(self.prolong_kernel, self.mesh.cell_set,
+                     self.uc.dat(op2.WRITE, self.cell_node_map),
+                     self.uf.dat(op2.READ, self.cell_node_map))
+
+        with self.uc.dat.vec_ro as xc:
+            xc.copy(y)
+
+        y.array_w[self.bc_nodes] = x.array_r[self.bc_nodes]
+
+    def view(self, pc, viewer=None):
+        super(FDMPC, self).view(pc, viewer)
```

Comment:
Is there anything else useful to print out here about the FDMPC (like quadrature degree, eta, ...)?